### PR TITLE
Feat/issue 82/extraction pipeline integration assets

### LINKED ISSUE #82: Use more accurate logic for borehole identification from swissgeol-boreholes-dataextraction repo
Follow-up for https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/310

Currently, this repository uses a few "heuristically" computed features for identifying borehole profiles, which are then passed to the classification model. There features are much less accurate than the much more advanced logic in the swissgeol-boreholes-dataextraction repository.

We should improve the accuracy of the detection of boreholes in geological documents, by running more of the borehole identification logic from the other repository. In this context, we mostly care about the question "is there a borehole profile on this page or not" as well as extracting the borehole name (in case there is a borehole profile). However, we want to avoid doing excessive computations for data that is not needed in this context, such as accurately computing the individual layers or extracting groundwater data. (To get a better pictures of this, it would be worthwhile to spend some time on creating a concept / diagrams / better documentation. Based on this, decide on what modules/classes/methods we want to expose from boreholes to assets.)

Use the borehole identification as an new input feature for the xgBoost model and retrain the model. Hopefully, there will be a very strong correlation between this feature and the target class "borehole". Later on, we could think about whether it makes sense to use the borehole identification also in different ways (independently of the xgBoost model).

### PR DESCRIPTION
**Pull request by letao**
Closes https://github.com/swisstopo/swissgeol-assets-dataextraction/issues/82 

This PR integrates borehole extraction logic from the swissgeol-boreholes-dataextraction repository to improve the accuracy of borehole profile detection in geological documents. The feature set expands from 17 to 23 features, adding 8 new borehole-specific features.

**Borehole features (8)**
  - Num Valid Borehole Descriptions
  - Num Strip Logs
  - Num Tables
  - Num Boreholes
  - Num Good Sidebars
  - Best Sidebar Score
  - Num of Vertical lines
  - Text Line Count
  
In [feature_engineering.py](https://github.com/swisstopo/swissgeol-assets-dataextraction/blob/feat/issue-82/extraction-pipeline-integration-assets/src/models/feature_engineering.py) `get_features() `and `get_features_from_page()` function are unified into a single function. 
 
Local `TextWord`, `TextLine` and `TextBlock` classes are substituted with imports from swissgeol_doc_processing package. The remaining classes are commented as deprectated. 

Extended `matching_params.yml` with affinity parameters and sidebar detection settings. Updated model configs to reflect. 

New Additions:

- [train_parallel.py](https://github.com/swisstopo/swissgeol-assets-dataextraction/blob/feat/issue-82/extraction-pipeline-integration-assets/src/models/treebased/train_parallel.py]) - Parallelized training script for XGBoost model.

### CODE REVIEW

**Comment by letao** [comment]
As a result of findings from this PR a follow-up issue has been created: https://github.com/swisstopo/swissgeol-assets-dataextraction/issues/91

File: config/randomforest_config.yml [hunk 1]
```diff
@@ -6,27 +6,38 @@ train_folder_path: "data/single_pages_splits/train"
 val_folder_path: "data/single_pages_splits/val"
 ground_truth_file_path: "data/gt_single_pages.json"
 
-# Feature names to track
+# Feature names to track (27 features total)
```
[the remark this answers is not in the document]
Comment (letao) [reply config/randomforest_config.yml:9]: Not Relevant since `randomforest` is currently not used and issue https://github.com/swisstopo/swissgeol-assets-dataextraction/issues/91 will address this

File: config/randomforest_config.yml [hunk 2]
```diff
@@ -6,27 +6,38 @@ train_folder_path: "data/single_pages_splits/train"
...
-  - Line score
+  - Line Score
+  # Geo profile and diagram features (4)
   - Num Geo Profile Keywords
   - Num Unit Keyword
-  - Y scale ok
-  - X scale ok
+  - Y Scale OK
+  - X Scale OK
+  # Borehole features (12)
```
[the remark this answers is not in the document]
Comment (letao) [reply config/randomforest_config.yml:29]: Not Relevant since `randomforest` is currently not used and issue https://github.com/swisstopo/swissgeol-assets-dataextraction/issues/91 will address this

File: config/randomforest_config.yml [hunk 3]
```diff
@@ -6,27 +6,38 @@ train_folder_path: "data/single_pages_splits/train"
...
+  - Num Strip Logs
+  - Num Tables
+  - Num Boreholes
+  - Num Sidebar Candidates
+  - Num Good Sidebars
+  - Best Sidebar Score
+  - Sidebar Types Found
+  - Average Sidebar Noise
+  - Num Long Or Horizontal Lines
+  - Text Line Count
```
[the remark this answers is not in the document]
Comment (letao) [reply config/randomforest_config.yml:40]: Not Relevant since `randomforest` is currently not used and issue https://github.com/swisstopo/swissgeol-assets-dataextraction/issues/91 will address this

File: config/randomforest_config.yml [hunk 4]
```diff
@@ -6,27 +6,38 @@ train_folder_path: "data/single_pages_splits/train"
 val_folder_path: "data/single_pages_splits/val"
 ground_truth_file_path: "data/gt_single_pages.json"
 
-# Feature names to track
+# Feature names to track (27 features total)
 feature_names:
+  # Text/word features (6)
   - Words Per Line
-  - Text zone Density
+  - Text Zone Density
```
[the remark this answers is not in the document]
Comment (letao) [reply config/randomforest_config.yml:13]: Not Relevant since `randomforest` is currently not used and issue https://github.com/swisstopo/swissgeol-assets-dataextraction/issues/91 will address this

File: config/table_detection_params.yml [hunk 5]
```diff
@@ -1,26 +1,29 @@
...
-  text_presence_weight: 0.01
+  # Line score for confidence calculation
+  line_scoring:
+    line_weights: 0.5
+    max_n_lines_bonus: 90
+
+  # Text bonus scoring for confidence calculation
+  text_scoring:
+    text_weights: 0.3
+    text_presence_weight: 0.01
```
[the remark this answers is not in the document]
Comment (letao) [reply config/table_detection_params.yml]: Not Relevant since `src/table_detection.py` that is used exclusively in `baseline classifier` which in turn is currently not used and issue https://github.com/swisstopo/swissgeol-assets-dataextraction/issues/91 will address this

File: config/matching_params.yml [hunk 6]
```diff
@@ -1,3 +1,73 @@
+block_line_ratio: 0.20
```
Comment (stijnvermeeren-swisstopo) [inline config/matching_params.yml]: Do we really need this file here? Can't we reuse the config from `swissgeol_doc_processing/config/matching_params.yml` in the `swissgeol-borehole-dataextraction` repo?

File: config/table_detection_params.yml [hunk 7]
```diff
@@ -1,26 +1,29 @@
-# Line length requirements for meaningful table detection
-min_line_length: 20
 
-# Angle requirements
-angle_tolerance: 5
+# Table detection parameters
```
Comment (stijnvermeeren-swisstopo) [inline config/table_detection_params.yml]: Same with this file, look identical to the one in the other repository.

File: tests/test_textline.py [hunk 8]
```diff
@@ -1,6 +1,5 @@
 import fitz
-
-from src.text_objects import TextLine, TextWord
+from swissgeol_doc_processing.text.textline import TextLine, TextWord
 
 
 def test_textline_with_valid_words():
```
Comment (stijnvermeeren-swisstopo) [inline tests/test_textline.py]: This just tests functionality that is imported from the other repository, If relevant, it should be unit tested there. I don't see a reason to keep including these unit tests here in this repository.

File: src/page_structure.py [hunk 9]
```diff
@@ -1,14 +1,21 @@
...
 from pydantic import BaseModel, ConfigDict, FieldSerializationInfo, field_serializer
+from swissgeol_doc_processing.geometry.geometry_dataclasses import Line
+from swissgeol_doc_processing.text.textblock import TextBlock
+from swissgeol_doc_processing.text.textline import TextLine, TextWord
 
-from src.geometric_objects import Line
 from src.page_classes import PageClasses
-from src.text_objects import TextBlock, TextLine, TextWord
+
+if TYPE_CHECKING:
```
Comment (stijnvermeeren-swisstopo) [inline src/page_structure.py]: What's the purpose of this? Why not always import `ExtractionContext`, since it is used in this file?

File: src/models/treebased/train_parallel.py [hunk 10]
```diff
@@ -0,0 +1,240 @@
+"""Parallelized version of train.py with multiprocessing for feature extraction."""
```
Comment (stijnvermeeren-swisstopo) [inline src/models/treebased/train_parallel.py:1]: Could you document in the README how to run this script, and what are the differences/advantages compared to the regular train.py script?

File: src/pdf_processor.py [hunk 11]
```diff
@@ -44,25 +50,44 @@ def __init__(self, classifier: Classifier):
...
+
         return PageContext(
             lines=lines,
             words=words,
             text_blocks=text_blocks,
             language=language,
             page_rect=page_rect,
             text_rect=text_rect,
-            geometric_lines=[],
+            geometric_lines=all_geometric_lines,
```
Comment (stijnvermeeren-swisstopo) [inline src/pdf_processor.py]: is there an advantage in keeping `geometric_lines` as a separate parameter, when the same value is contained in the extraction context anyway?

File: src/pdf_processor.py [hunk 12]
```diff
@@ -44,25 +50,44 @@ def __init__(self, classifier: Classifier):
...
+        table_structures = detect_table_structures(page, long_or_horizontal_lines, lines, table_detection_params)
+
+        extraction_context = ExtractionContext(
+            text_lines=lines,
+            long_or_horizontal_lines=long_or_horizontal_lines,
+            all_geometric_lines=all_geometric_lines,
+            strip_logs=strip_logs,
+            table_structures=table_structures,
+            language=language,
+        )
```
Comment (stijnvermeeren-swisstopo) [inline src/pdf_processor.py]: Can't we use `ExtractionContext.from_page` here?

**Review by stijnvermeeren-swisstopo — CHANGES_REQUESTED** [review]

File: main.py [hunk 13]
```diff
@@ -183,7 +188,8 @@ def main(
     """
     input_path = Path(input_path)
     ground_truth_path = Path(ground_truth_path) if ground_truth_path else None
-    matching_params = read_params("config/matching_params.yml")
+    matching_params = read_params("config/local_matching_params.yml")
```
[the remark this answers is not in the document]
Comment (letao) [reply main.py:191]: To be resolved in https://github.com/swisstopo/swissgeol-assets-dataextraction/issues/91

File: src/scripts/download_from_s3.py [hunk 14]
```diff
@@ -34,7 +34,7 @@
 OUTPUT_DIR = DATA_DIR / "single_pages_new"
 OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
 
-MATCHING_PARAMS = read_params(REPO_ROOT / "config" / "matching_params.yml")
+MATCHING_PARAMS = read_params(REPO_ROOT / "config" / "local_matching_params.yml")
```
[the remark this answers is not in the document]
Comment (letao) [reply src/scripts/download_from_s3.py:37]: To be resolved in https://github.com/swisstopo/swissgeol-assets-dataextraction/issues/91

File: src/classifiers/baseline_classifier.py [hunk 15]
```diff
@@ -101,7 +101,11 @@ def _extract_geometric_lines(self, context: PageContext, page: pymupdf.Page):
         Returns:
             None: The function modifies the context in place by setting the geometric_lines attribute.
         """
-        geometric_lines = extract_geometric_lines(page)
+        # Use cached geometric lines if available from ExtractionContext
+        if context.extraction_context is not None:
+            geometric_lines = list(context.extraction_context.all_geometric_lines)
+        else:
+            geometric_lines = extract_geometric_lines(page)
```
[the remark this answers is not in the document]
Comment (letao) [reply src/classifiers/baseline_classifier.py:108]: To be resolved in https://github.com/swisstopo/swissgeol-assets-dataextraction/issues/91

File: src/pdf_processor.py [hunk 16]
```diff
@@ -42,27 +44,37 @@ def __init__(self, classifier: Classifier):
...
+            page, line_detection_params, striplog_detection_params, table_detection_params
+        )
+
+        extraction_context.language = language
+
         is_digital = is_digitally_born(page)
         words = extract_words(page, page_number)
-        lines = create_text_lines(page, page_number)
+        lines = extraction_context.text_lines
         text_blocks = create_text_blocks(lines)
```
[the remark this answers is not in the document]
Comment (letao) [reply src/pdf_processor.py:61]: Currently there are no negative impacts on Treebased Classifier when switching to swissgeol_doc_processing. Since full swith requires rewriting of unit tests this will be resolved in https://github.com/swisstopo/swissgeol-assets-dataextraction/issues/91

File: src/models/feature_engineering.py [hunk 17]
```diff
@@ -140,14 +165,72 @@ def compute_text_features(
...
+        name_detection_params: Optional parameters for name detection.
+        extraction_context: Optional ExtractionContext for caching intermediate results.
+
+    Returns:
+        List of 8 feature values in consistent order for classification.
+    """
+    if line_detection_params is None:
+        line_detection_params = read_params("line_detection_params.yml")
+    if name_detection_params is None:
+        name_detection_params = read_params("name_detection_params.yml")
```
[the remark this answers is not in the document]
Comment (letao) [reply src/models/feature_engineering.py:203]: To be resolved in https://github.com/swisstopo/swissgeol-assets-dataextraction/issues/91

**Review by stijnvermeeren-swisstopo — APPROVED** [review]
LGTM!
